# Manipulating DataFrames

Regardless of the original data source, once you have data loaded into a DataFrame, you gain the ability to manipulate your data. For example, you can do the following:

  - Aggregate a column (max, min, sum, etc)
  - Reshape:
    - Sort
    - Drop columns
    - `melt()` and `pivot()`
  - Select rows, using the following:
    - Logical criteria
    - `head()` and `tail()`
    - `iloc`
  - Select columns
  - Handle missing data with `dropna()` and `fillna()`
  - Make new columns
  - Combine datasets with `merge()`
  - Group data
    
To explore this, read in a data table consisting of the salaries and personal statistics of major league baseball players.

In [1]:
import pandas as pd

players_df = pd.read_csv("https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/players.csv")

First, examine your DataFrame using some of the techniques that you looked at in the previous lesson.

In [2]:
players_df.shape

(26428, 14)

You can refer to any individual column of the DataFrame by placing it inside square brackets `[]` and quotations `''`.

In [3]:
# Total salary
players_df['salary'].sum()

55119136756

In [4]:
# Max height
players_df['height'].max()

83

If you don't specify a given column or columns, the method will apply to the entire DataFrame:

In [5]:
players_df.describe()

birthyear    deathyear        weight        height        yearid  \
count  26428.000000   492.000000  26428.000000  26428.000000  26428.000000   
mean    1971.389133  2008.833333    199.022136     73.509006   2000.878727   
std        9.679736     7.188803     22.631696      2.284665      8.909314   
min     1925.000000  1989.000000    140.000000     66.000000   1985.000000   
25%     1964.000000  2006.000000    185.000000     72.000000   1994.000000   
50%     1971.000000  2011.000000    195.000000     74.000000   2001.000000   
75%     1979.000000  2015.000000    215.000000     75.000000   2009.000000   
max     1995.000000  2018.000000    315.000000     83.000000   2016.000000   

             salary  
count  2.642800e+04  
mean   2.085634e+06  
std    3.455348e+06  
min    0.000000e+00  
25%    2.947020e+05  
50%    5.500000e+05  
75%    2.350000e+06  
max    3.300000e+07

In [6]:
players_df.columns

Index(['playerid', 'birthyear', 'birthcountry', 'deathyear', 'namefirst',
       'namelast', 'weight', 'height', 'bats', 'throws', 'yearid', 'teamid',
       'lgid', 'salary'],
      dtype='object')

In [7]:
players_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26428 entries, 0 to 26427
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   playerid      26428 non-null  object 
 1   birthyear     26428 non-null  int64  
 2   birthcountry  26428 non-null  object 
 3   deathyear     492 non-null    float64
 4   namefirst     26428 non-null  object 
 5   namelast      26428 non-null  object 
 6   weight        26428 non-null  int64  
 7   height        26428 non-null  int64  
 8   bats          26428 non-null  object 
 9   throws        26428 non-null  object 
 10  yearid        26428 non-null  int64  
 11  teamid        26428 non-null  object 
 12  lgid          26428 non-null  object 
 13  salary        26428 non-null  int64  
dtypes: float64(1), int64(5), object(8)
memory usage: 2.8+ MB


In [8]:
players_df

playerid  birthyear birthcountry  deathyear namefirst   namelast  \
0      barkele01       1955          USA        NaN       Len     Barker   
1      bedrost01       1957          USA        NaN     Steve  Bedrosian   
2      benedbr01       1955          USA        NaN     Bruce   Benedict   
3       campri01       1953          USA     2013.0      Rick       Camp   
4      ceronri01       1954          USA        NaN      Rick     Cerone   
...          ...        ...          ...        ...       ...        ...   
26423  strasst01       1988          USA        NaN   Stephen  Strasburg   
26424  taylomi02       1991          USA        NaN   Michael     Taylor   
26425  treinbl01       1988          USA        NaN     Blake    Treinen   
26426  werthja01       1979          USA        NaN    Jayson      Werth   
26427  zimmery01       1984          USA        NaN      Ryan  Zimmerman   

       weight  height bats throws  yearid teamid lgid    salary  
0         225      77    R      R    1985    ATL   NL    870000  
1         200      75    R      R    1985    ATL   NL    550000  
2         175      73    R      R    1985    ATL   NL    545000  
3         195      73    R      R    1985    ATL   NL    633333  
4         192      71    R      R    1985    ATL   NL    625000  
...       ...     ...  ...    ...     ...    ...  ...       ...  
26423     235      76    R      R    2016    WAS   NL  10400000  
26424     210      75    R      R    2016    WAS   NL    524000  
26425     225      77    R      R    2016    WAS   NL    524900  
26426     235      77    R      R    2016    WAS   NL  21733615  
26427     225      75    R      R    2016    WAS   NL  14000000  

[26428 rows x 14 columns]

### Missing values

`NaN` in pandas represents a null value, meaning that a value is missing. Note that a null value doesn't necessarily mean that the number is `0`. It could be missing due to something like a recording error, its inapplicability, or sampling bias. In later modules, you will take a closer look at how to diagnose missing values.  

The `isnull()` method returns a boolean result indicating whether each value in a DataFrame is missing. `True` equates to a missing value.

In [9]:
players_df.isnull()

playerid  birthyear  birthcountry  deathyear  namefirst  namelast  \
0         False      False         False       True      False     False   
1         False      False         False       True      False     False   
2         False      False         False       True      False     False   
3         False      False         False      False      False     False   
4         False      False         False       True      False     False   
...         ...        ...           ...        ...        ...       ...   
26423     False      False         False       True      False     False   
26424     False      False         False       True      False     False   
26425     False      False         False       True      False     False   
26426     False      False         False       True      False     False   
26427     False      False         False       True      False     False   

       weight  height   bats  throws  yearid  teamid   lgid  salary  
0       False   False  False   False   False   False  False   False  
1       False   False  False   False   False   False  False   False  
2       False   False  False   False   False   False  False   False  
3       False   False  False   False   False   False  False   False  
4       False   False  False   False   False   False  False   False  
...       ...     ...    ...     ...     ...     ...    ...     ...  
26423   False   False  False   False   False   False  False   False  
26424   False   False  False   False   False   False  False   False  
26425   False   False  False   False   False   False  False   False  
26426   False   False  False   False   False   False  False   False  
26427   False   False  False   False   False   False  False   False  

[26428 rows x 14 columns]

`isna()` can also be used to do the same thing:

In [10]:
players_df.isna()

playerid  birthyear  birthcountry  deathyear  namefirst  namelast  \
0         False      False         False       True      False     False   
1         False      False         False       True      False     False   
2         False      False         False       True      False     False   
3         False      False         False      False      False     False   
4         False      False         False       True      False     False   
...         ...        ...           ...        ...        ...       ...   
26423     False      False         False       True      False     False   
26424     False      False         False       True      False     False   
26425     False      False         False       True      False     False   
26426     False      False         False       True      False     False   
26427     False      False         False       True      False     False   

       weight  height   bats  throws  yearid  teamid   lgid  salary  
0       False   False  False   False   False   False  False   False  
1       False   False  False   False   False   False  False   False  
2       False   False  False   False   False   False  False   False  
3       False   False  False   False   False   False  False   False  
4       False   False  False   False   False   False  False   False  
...       ...     ...    ...     ...     ...     ...    ...     ...  
26423   False   False  False   False   False   False  False   False  
26424   False   False  False   False   False   False  False   False  
26425   False   False  False   False   False   False  False   False  
26426   False   False  False   False   False   False  False   False  
26427   False   False  False   False   False   False  False   False  

[26428 rows x 14 columns]

Immediately, you can see that there are quite a few missing values in the *deathyear* field. It makes sense that there are missing values here—if a player is still living, then their death year is indeed unknown. Making this connection shows the importance of understanding how data is collected; the data itself may not provide the answers.

Should you want to exclude all records with a missing value, you can call the `dropna()` method. By default, this will drop any row that includes a missing value in any field. There are many other ways to drop missing values in pandas; to learn more, check the [official documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html).

In [11]:
players_df_filtered = players_df.dropna()
players_df_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 492 entries, 3 to 25988
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   playerid      492 non-null    object 
 1   birthyear     492 non-null    int64  
 2   birthcountry  492 non-null    object 
 3   deathyear     492 non-null    float64
 4   namefirst     492 non-null    object 
 5   namelast      492 non-null    object 
 6   weight        492 non-null    int64  
 7   height        492 non-null    int64  
 8   bats          492 non-null    object 
 9   throws        492 non-null    object 
 10  yearid        492 non-null    int64  
 11  teamid        492 non-null    object 
 12  lgid          492 non-null    object 
 13  salary        492 non-null    int64  
dtypes: float64(1), int64(5), object(8)
memory usage: 57.7+ KB


Your DataFrame now contains only 492 rows—all rows where the player's death year is unknown have been dropped. 

But what if instead of dropping these rows, you want to fill in the missing values with something other than `NaN`? You can do this using the `fillna()` method. 

This can be useful for making a DataFrame more legible, or for assigning a special value or character to nulls. 


In [12]:
players_df.fillna('To be determined')

playerid  birthyear birthcountry         deathyear namefirst  \
0      barkele01       1955          USA  To be determined       Len   
1      bedrost01       1957          USA  To be determined     Steve   
2      benedbr01       1955          USA  To be determined     Bruce   
3       campri01       1953          USA            2013.0      Rick   
4      ceronri01       1954          USA  To be determined      Rick   
...          ...        ...          ...               ...       ...   
26423  strasst01       1988          USA  To be determined   Stephen   
26424  taylomi02       1991          USA  To be determined   Michael   
26425  treinbl01       1988          USA  To be determined     Blake   
26426  werthja01       1979          USA  To be determined    Jayson   
26427  zimmery01       1984          USA  To be determined      Ryan   

        namelast  weight  height bats throws  yearid teamid lgid    salary  
0         Barker     225      77    R      R    1985    ATL   NL    870000  
1      Bedrosian     200      75    R      R    1985    ATL   NL    550000  
2       Benedict     175      73    R      R    1985    ATL   NL    545000  
3           Camp     195      73    R      R    1985    ATL   NL    633333  
4         Cerone     192      71    R      R    1985    ATL   NL    625000  
...          ...     ...     ...  ...    ...     ...    ...  ...       ...  
26423  Strasburg     235      76    R      R    2016    WAS   NL  10400000  
26424     Taylor     210      75    R      R    2016    WAS   NL    524000  
26425    Treinen     225      77    R      R    2016    WAS   NL    524900  
26426      Werth     235      77    R      R    2016    WAS   NL  21733615  
26427  Zimmerman     225      75    R      R    2016    WAS   NL  14000000  

[26428 rows x 14 columns]

## Add new columns

Sometimes, you'll need to add new columns by deriving values from calculations involving other columns. For example, if you had data on the weight and height of all players on a team, you could compute a new column for body mass index (BMI) using the following formula:

$$ bmi = \frac{weight * 703}{height^2} $$

To create a column of values calculated from the values in other columns, use the `assign()` method of the DataFrame.

In [13]:
players_df = players_df.assign(bmi = (703 * players_df.weight) / (players_df.height**2))
players_df.head()

playerid  birthyear birthcountry  deathyear namefirst   namelast  weight  \
0  barkele01       1955          USA        NaN       Len     Barker     225   
1  bedrost01       1957          USA        NaN     Steve  Bedrosian     200   
2  benedbr01       1955          USA        NaN     Bruce   Benedict     175   
3   campri01       1953          USA     2013.0      Rick       Camp     195   
4  ceronri01       1954          USA        NaN      Rick     Cerone     192   

   height bats throws  yearid teamid lgid  salary        bmi  
0      77    R      R    1985    ATL   NL  870000  26.678192  
1      75    R      R    1985    ATL   NL  550000  24.995556  
2      73    R      R    1985    ATL   NL  545000  23.085945  
3      73    R      R    1985    ATL   NL  633333  25.724339  
4      71    R      R    1985    ATL   NL  625000  26.775640

Or you could simply add a new column with some default:

In [14]:
players_df['X'] = 0
players_df.head()

playerid  birthyear birthcountry  deathyear namefirst   namelast  weight  \
0  barkele01       1955          USA        NaN       Len     Barker     225   
1  bedrost01       1957          USA        NaN     Steve  Bedrosian     200   
2  benedbr01       1955          USA        NaN     Bruce   Benedict     175   
3   campri01       1953          USA     2013.0      Rick       Camp     195   
4  ceronri01       1954          USA        NaN      Rick     Cerone     192   

   height bats throws  yearid teamid lgid  salary        bmi  X  
0      77    R      R    1985    ATL   NL  870000  26.678192  0  
1      75    R      R    1985    ATL   NL  550000  24.995556  0  
2      73    R      R    1985    ATL   NL  545000  23.085945  0  
3      73    R      R    1985    ATL   NL  633333  25.724339  0  
4      71    R      R    1985    ATL   NL  625000  26.775640  0

Both of those options are row-wise operators.

## Deleting columns

To remove a column, use the `drop()` method. This method can be used to drop both rows and columns, so the axes have to be specified. Try it out by removing the new column that you just created. By default, the `drop()` method does not actually modify the DataFrame; rather, it creates a new DataFrame with the changes, leaving the existing one untouched. This can be useful in many cases, but you need to be aware of that behavior. 

As an alternative, you can set the `inplace` parameter to `True` to make the operation work on the existing DataFrame. The following code snippet first drops a column and assigns the new modified DataFrame to a new variable named *players_df_changed*. The original DataFrame, *players_df*, still exists. Next, the original DataFrame is modified by dropping the same column while passing the `inplace` parameter set to `True`.

In [15]:
# Delete the `X` column and store the new DataFrame in a new variable
players_df_changed = players_df.drop(['X'], axis=1)

print('The columns of the new DataFrame')
print(players_df_changed.columns)

print('The columns of the original DataFrame')
print(players_df.columns)

# Alternatively, modify the existing DataFrame by setting the `inplace` parameter to `True`
players_df.drop(['X'], axis=1, inplace=True)
print('The columns of the DataFrame after using the `inplace` parameter for deletion')
print(players_df.columns)


The columns of the new DataFrame
Index(['playerid', 'birthyear', 'birthcountry', 'deathyear', 'namefirst',
       'namelast', 'weight', 'height', 'bats', 'throws', 'yearid', 'teamid',
       'lgid', 'salary', 'bmi'],
      dtype='object')
The columns of the original DataFrame
Index(['playerid', 'birthyear', 'birthcountry', 'deathyear', 'namefirst',
       'namelast', 'weight', 'height', 'bats', 'throws', 'yearid', 'teamid',
       'lgid', 'salary', 'bmi', 'X'],
      dtype='object')
The columns of the DataFrame after using the `inplace` parameter for deletion
Index(['playerid', 'birthyear', 'birthcountry', 'deathyear', 'namefirst',
       'namelast', 'weight', 'height', 'bats', 'throws', 'yearid', 'teamid',
       'lgid', 'salary', 'bmi'],
      dtype='object')


## Selecting rows

When you're exploring data in pandas, one thing that you'll need to do constantly is select a subset of rows based on some criterion. Suppose, for example, that you need to see the data for just a single year. Or maybe you need to select some rows by numbers.

### Selecting by number

You have already seen that you can get the first `n` rows or the last `n` rows of a DataFrame using `head(n)` and `tail(n)`, respectively. But what if you wanted to select rows 20-25? To select rows based on position, use `iloc`.  



In [16]:
# Select the first row
players_df.iloc[0]

playerid        barkele01
birthyear            1955
birthcountry          USA
deathyear             NaN
namefirst             Len
namelast           Barker
weight                225
height                 77
bats                    R
throws                  R
yearid               1985
teamid                ATL
lgid                   NL
salary             870000
bmi             26.678192
Name: 0, dtype: object

The `iloc` selector can take inputs in several different ways:

 - An integer
 - A list of integers
 - A slice object
 - A boolean array
 
When a single row is selected, the result is a *Series*—a single column of values indexed with the column names. When a list of integers is provided, the result is a DataFrame instead.

In [17]:
# Select the first row, this time with a list of integers
players_df.iloc[[0]]

playerid  birthyear birthcountry  deathyear namefirst namelast  weight  \
0  barkele01       1955          USA        NaN       Len   Barker     225   

   height bats throws  yearid teamid lgid  salary        bmi  
0      77    R      R    1985    ATL   NL  870000  26.678192

To select more than one row, just list the row numbers.

In [18]:
players_df.iloc[[0, 1, 5, 8, 10]]

playerid  birthyear birthcountry  deathyear namefirst   namelast  weight  \
0   barkele01       1955          USA        NaN       Len     Barker     225   
1   bedrost01       1957          USA        NaN     Steve  Bedrosian     200   
5   chambch01       1948          USA        NaN     Chris  Chambliss     195   
8   garbege01       1947          USA        NaN      Gene     Garber     175   
10  hornebo01       1957          USA        NaN       Bob     Horner     195   

    height bats throws  yearid teamid lgid   salary        bmi  
0       77    R      R    1985    ATL   NL   870000  26.678192  
1       75    R      R    1985    ATL   NL   550000  24.995556  
5       73    L      R    1985    ATL   NL   800000  25.724339  
8       70    R      R    1985    ATL   NL   772000  25.107143  
10      73    R      R    1985    ATL   NL  1500000  25.724339

To select multiple rows with a sequential index, use a slice object. In Python, a slice object takes the following form:

```
[start:end]
```

Here, `start` and `end` are integers. It is basically a list of numbers from `start` to `end-1`. For example, `[2:5]` is the same as `[2, 3, 4]`.



In [19]:
# Select rows 5-10
players_df.iloc[5:11]

playerid  birthyear birthcountry  deathyear namefirst   namelast  weight  \
5   chambch01       1948          USA        NaN     Chris  Chambliss     195   
6   dedmoje01       1960          USA        NaN      Jeff     Dedmon     200   
7   forstte01       1952          USA        NaN     Terry    Forster     200   
8   garbege01       1947          USA        NaN      Gene     Garber     175   
9   harpete01       1955          USA        NaN     Terry     Harper     195   
10  hornebo01       1957          USA        NaN       Bob     Horner     195   

    height bats throws  yearid teamid lgid   salary        bmi  
5       73    L      R    1985    ATL   NL   800000  25.724339  
6       74    L      R    1985    ATL   NL   150000  25.675676  
7       75    L      L    1985    ATL   NL   483333  24.995556  
8       70    R      R    1985    ATL   NL   772000  25.107143  
9       76    R      R    1985    ATL   NL   250000  23.733553  
10      73    R      R    1985    ATL   NL  1500000  25.724339

Alternatively, `iloc` may take a list of booleans that is the same length as the index and only return the rows that are `True`. For example, if the list `[False, True, True, False]` is used to select from a DataFrame with four rows, it will skip the first row, take the second and third rows, and skip the last row.

Your DataFrame has over 26,000 rows, and typing a list of 26,000 boolean values is just not practical. But you could generate a list based on some value in the DataFrame. For example, suppose that you want to get a list of all players who weigh over 200 pounds. You could generate a list of booleans like this:

In [20]:
# Create a Series of booleans
over_200 = players_df['weight'] > 200
over_200.head(10)

0     True
1    False
2    False
3    False
4    False
5    False
6    False
7    False
8    False
9    False
Name: weight, dtype: bool

The *over_200* Series now contains a boolean value for each row in the DataFrame. You can see how many `True` values and how many `False` values there are with the `value_counts()` method.

In [21]:
over_200.value_counts()

False    15829
True     10599
Name: weight, dtype: int64

To select only the players corresponding to a `True` value, you can pass the array of booleans to `iloc`.

In [22]:
players_df.iloc[over_200.values]


playerid  birthyear birthcountry  deathyear namefirst   namelast  \
0      barkele01       1955          USA        NaN       Len     Barker   
15     murphda05       1956          USA        NaN      Dale     Murphy   
24     davisst02       1961          USA        NaN     Storm      Davis   
29     grosswa01       1952          USA        NaN     Wayne      Gross   
39     roeniga01       1954          USA        NaN      Gary   Roenicke   
...          ...        ...          ...        ...       ...        ...   
26423  strasst01       1988          USA        NaN   Stephen  Strasburg   
26424  taylomi02       1991          USA        NaN   Michael     Taylor   
26425  treinbl01       1988          USA        NaN     Blake    Treinen   
26426  werthja01       1979          USA        NaN    Jayson      Werth   
26427  zimmery01       1984          USA        NaN      Ryan  Zimmerman   

       weight  height bats throws  yearid teamid lgid    salary        bmi  
0         225      77    R      R    1985    ATL   NL    870000  26.678192  
15        210      76    R      R    1985    ATL   NL   1625000  25.559211  
24        207      76    R      R    1985    BAL   AL    437500  25.194079  
29        210      74    L      R    1985    BAL   AL    483333  26.959459  
39        205      75    R      R    1985    BAL   AL    558333  25.620444  
...       ...     ...  ...    ...     ...    ...  ...       ...        ...  
26423     235      76    R      R    2016    WAS   NL  10400000  28.601974  
26424     210      75    R      R    2016    WAS   NL    524000  26.245333  
26425     225      77    R      R    2016    WAS   NL    524900  26.678192  
26426     235      77    R      R    2016    WAS   NL  21733615  27.863889  
26427     225      75    R      R    2016    WAS   NL  14000000  28.120000  

[10599 rows x 15 columns]

The result is a DataFrame with 10,599 rows. Here is another example with multiple conditions: select all players over 200 pounds who bat left-handed and throw right-handed.

In [23]:
over_200_L_R = (players_df['weight'] > 200) & (players_df['bats'] == 'L') & (players_df['throws'] == 'R')
players_df.iloc[over_200_L_R.values]

playerid  birthyear birthcountry  deathyear namefirst   namelast  \
29     grosswa01       1952          USA        NaN     Wayne      Gross   
40     sheetla01       1959          USA        NaN     Larry     Sheets   
55     gedmari01       1959          USA        NaN      Rich     Gedman   
117    walkegr01       1959          USA        NaN      Greg     Walker   
137    sutclri01       1956          USA        NaN      Rick  Sutcliffe   
...          ...        ...          ...        ...       ...        ...   
26394  saundmi01       1986          CAN        NaN   Michael   Saunders   
26399  tholejo01       1986          USA        NaN      Josh      Thole   
26404   drewst01       1983          USA        NaN   Stephen       Drew   
26407  harpebr03       1992          USA        NaN     Bryce     Harper   
26411  murphda08       1985          USA        NaN    Daniel     Murphy   

       weight  height bats throws  yearid teamid lgid   salary        bmi  
29        210      74    L      R    1985    BAL   AL   483333  26.959459  
40        217      75    L      R    1985    BAL   AL    60000  27.120178  
55        210      72    L      R    1985    BOS   AL   477500  28.478009  
117       205      75    L      R    1985    CHA   AL   195000  25.620444  
137       215      79    L      R    1985    CHN   NL  1260000  24.218074  
...       ...     ...  ...    ...     ...    ...  ...      ...        ...  
26394     225      76    L      R    2016    TOR   AL  2900000  27.384868  
26399     205      73    L      R    2016    TOR   AL   800000  27.043535  
26404     220      72    L      R    2016    WAS   NL  3000000  29.834105  
26407     215      75    L      R    2016    WAS   NL  5000000  26.870222  
26411     215      73    L      R    2016    WAS   NL  8000000  28.362732  

[1072 rows x 15 columns]

### Selecting on both axes

Sometimes you want to select only some columns of the DataFrame. The `iloc` selector allows that with a second value representing the second axis. Like the rows, the columns are numbered from `0`. So, to select the first name, last name, weight, height, bats, and throws columns for the first 10 rows of the DataFrame, you would need to specify a slice of `[4:10]` for the columns.



In [24]:
# Rows 0-9 and columns 5-10
players_df.iloc[0:10, 4:10]

namefirst   namelast  weight  height bats throws
0       Len     Barker     225      77    R      R
1     Steve  Bedrosian     200      75    R      R
2     Bruce   Benedict     175      73    R      R
3      Rick       Camp     195      73    R      R
4      Rick     Cerone     192      71    R      R
5     Chris  Chambliss     195      73    L      R
6      Jeff     Dedmon     200      74    L      R
7     Terry    Forster     200      75    L      L
8      Gene     Garber     175      70    R      R
9     Terry     Harper     195      76    R      R

To make the rest of this discussion a little easier to illustrate, select a subset of the data. Select all players that played in the team CLE in 2015. And limit yourself to the following columns: *playerid*, *birthyear*, *namefirst*, *namelast*, *weight*, *height*, *bats*, *throws*, and *salary*.

In [25]:
# Create a Series of booleans for the row selection
cle_options = (players_df['teamid'] == 'CLE') & (players_df['yearid'] == 2015)
cle_2015 = players_df.iloc[cle_options.values, [0, 1, 4, 5, 6, 7, 8, 9, 13]]
cle_2015

playerid  birthyear namefirst     namelast  weight  height bats  \
24953  adamsau01       1986    Austin        Adams     200      71    R   
24954  allenco01       1988      Cody        Allen     210      73    R   
24955  atchisc01       1976     Scott     Atchison     200      74    R   
24956  avilemi01       1981      Mike       Aviles     205      70    R   
24957  bauertr01       1991    Trevor        Bauer     190      73    R   
24958  bournmi01       1982   Michael        Bourn     190      71    L   
24959  brantmi02       1987   Michael     Brantley     200      74    L   
24960  carraca01       1987    Carlos     Carrasco     212      75    R   
24961  chiselo01       1988    Lonnie   Chisenhall     190      74    L   
24962  crockky01       1991      Kyle     Crockett     175      74    L   
24963  floydga01       1983     Gavin        Floyd     245      76    R   
24964  gomesya01       1987       Yan        Gomes     215      74    R   
24965  hagadni01       1986      Nick     Hagadone     230      77    L   
24966  housetj01       1989     T. J.        House     205      73    R   
24967  kipnija01       1987     Jason       Kipnis     195      71    L   
24968  klubeco01       1986     Corey       Kluber     215      76    R   
24969  mcallza01       1987      Zach   McAllister     240      78    R   
24970   mossbr01       1983   Brandon         Moss     210      73    L   
24971  murphda07       1981     David       Murphy     210      75    L   
24972  perezro02       1988   Roberto        Perez     220      71    R   
24973  raburry01       1981      Ryan       Raburn     185      72    R   
24974  ramirjo01       1992      Jose      Ramirez     165      69    B   
24975  rzepcma01       1985      Marc  Rzepczynski     220      74    L   
24976  santaca01       1986    Carlos      Santana     210      71    B   
24977   shawbr01       1987     Bryan         Shaw     220      73    B   
24978  swarzan01       1985   Anthony      Swarzak     215      76    R   
24979  swishni01       1980      Nick      Swisher     195      72    B   
24980  tomlijo01       1984      Josh       Tomlin     190      73    R   
24981  walteza01       1989      Zach      Walters     210      74    B   

      throws    salary  
24953      R    507700  
24954      R    547100  
24955      R    900000  
24956      R   3500000  
24957      R   1940000  
24958      R  13500000  
24959      L   5875000  
24960      R   2337500  
24961      R   2250000  
24962      L    510900  
24963      R   4000000  
24964      R   1000000  
24965      L    517100  
24966      L    511900  
24967      R   4166666  
24968      R   1200000  
24969      R    520400  
24970      R   6500000  
24971      L   6000000  
24972      R    508600  
24973      R   2500000  
24974      R    511300  
24975      L   2400000  
24976      R   6000000  
24977      R   1550000  
24978      R    900000  
24979      L  15000000  
24980      R   1500000  
24981      R    509600

## Sorting

To sort a DataFrame, use `sort_values()`. It's possible to sort by the values in one or more columns, either in ascending order or descending order. The `sort_values()` method takes several parameters:

 - `by`: Either the name of a single column or a list of names of columns
 - `axis`: Either `0` to sort rows or `1` to sort columns; defaults to `0`
 - `ascending`: `True` (or `False` for `descending`); defaults to `True`
 - `inplace`: `True` to sort in place and modify the DataFrame, and `False` to create a new DataFrame; defaults to `False`
 - `na_position`: Where to put `NaN` values, either `first` or `last`; defaults to `last`

In [26]:
# Sort the players by weight from lowest value to highest value (ascending)
# Accept all other defaults
weight_asc = cle_2015.sort_values(by='weight')
weight_asc

playerid  birthyear namefirst     namelast  weight  height bats  \
24974  ramirjo01       1992      Jose      Ramirez     165      69    B   
24962  crockky01       1991      Kyle     Crockett     175      74    L   
24973  raburry01       1981      Ryan       Raburn     185      72    R   
24957  bauertr01       1991    Trevor        Bauer     190      73    R   
24958  bournmi01       1982   Michael        Bourn     190      71    L   
24980  tomlijo01       1984      Josh       Tomlin     190      73    R   
24961  chiselo01       1988    Lonnie   Chisenhall     190      74    L   
24967  kipnija01       1987     Jason       Kipnis     195      71    L   
24979  swishni01       1980      Nick      Swisher     195      72    B   
24953  adamsau01       1986    Austin        Adams     200      71    R   
24959  brantmi02       1987   Michael     Brantley     200      74    L   
24955  atchisc01       1976     Scott     Atchison     200      74    R   
24966  housetj01       1989     T. J.        House     205      73    R   
24956  avilemi01       1981      Mike       Aviles     205      70    R   
24981  walteza01       1989      Zach      Walters     210      74    B   
24954  allenco01       1988      Cody        Allen     210      73    R   
24970   mossbr01       1983   Brandon         Moss     210      73    L   
24971  murphda07       1981     David       Murphy     210      75    L   
24976  santaca01       1986    Carlos      Santana     210      71    B   
24960  carraca01       1987    Carlos     Carrasco     212      75    R   
24964  gomesya01       1987       Yan        Gomes     215      74    R   
24968  klubeco01       1986     Corey       Kluber     215      76    R   
24978  swarzan01       1985   Anthony      Swarzak     215      76    R   
24977   shawbr01       1987     Bryan         Shaw     220      73    B   
24972  perezro02       1988   Roberto        Perez     220      71    R   
24975  rzepcma01       1985      Marc  Rzepczynski     220      74    L   
24965  hagadni01       1986      Nick     Hagadone     230      77    L   
24969  mcallza01       1987      Zach   McAllister     240      78    R   
24963  floydga01       1983     Gavin        Floyd     245      76    R   

      throws    salary  
24974      R    511300  
24962      L    510900  
24973      R   2500000  
24957      R   1940000  
24958      R  13500000  
24980      R   1500000  
24961      R   2250000  
24967      R   4166666  
24979      L  15000000  
24953      R    507700  
24959      L   5875000  
24955      R    900000  
24966      L    511900  
24956      R   3500000  
24981      R    509600  
24954      R    547100  
24970      R   6500000  
24971      L   6000000  
24976      R   6000000  
24960      R   2337500  
24964      R   1000000  
24968      R   1200000  
24978      R    900000  
24977      R   1550000  
24972      R    508600  
24975      L   2400000  
24965      L    517100  
24969      R    520400  
24963      R   4000000

Suppose that you want to sort by weight and then by height. Look at the sorted DataFrame above. Notice that sometimes players with the same weight may have their heights in the wrong order. To fix this, you can provide a list of columns to sort by.

In [27]:
weight_height_asc = cle_2015.sort_values(by=['weight', 'height'])
weight_height_asc

playerid  birthyear namefirst     namelast  weight  height bats  \
24974  ramirjo01       1992      Jose      Ramirez     165      69    B   
24962  crockky01       1991      Kyle     Crockett     175      74    L   
24973  raburry01       1981      Ryan       Raburn     185      72    R   
24958  bournmi01       1982   Michael        Bourn     190      71    L   
24957  bauertr01       1991    Trevor        Bauer     190      73    R   
24980  tomlijo01       1984      Josh       Tomlin     190      73    R   
24961  chiselo01       1988    Lonnie   Chisenhall     190      74    L   
24967  kipnija01       1987     Jason       Kipnis     195      71    L   
24979  swishni01       1980      Nick      Swisher     195      72    B   
24953  adamsau01       1986    Austin        Adams     200      71    R   
24955  atchisc01       1976     Scott     Atchison     200      74    R   
24959  brantmi02       1987   Michael     Brantley     200      74    L   
24956  avilemi01       1981      Mike       Aviles     205      70    R   
24966  housetj01       1989     T. J.        House     205      73    R   
24976  santaca01       1986    Carlos      Santana     210      71    B   
24954  allenco01       1988      Cody        Allen     210      73    R   
24970   mossbr01       1983   Brandon         Moss     210      73    L   
24981  walteza01       1989      Zach      Walters     210      74    B   
24971  murphda07       1981     David       Murphy     210      75    L   
24960  carraca01       1987    Carlos     Carrasco     212      75    R   
24964  gomesya01       1987       Yan        Gomes     215      74    R   
24968  klubeco01       1986     Corey       Kluber     215      76    R   
24978  swarzan01       1985   Anthony      Swarzak     215      76    R   
24972  perezro02       1988   Roberto        Perez     220      71    R   
24977   shawbr01       1987     Bryan         Shaw     220      73    B   
24975  rzepcma01       1985      Marc  Rzepczynski     220      74    L   
24965  hagadni01       1986      Nick     Hagadone     230      77    L   
24969  mcallza01       1987      Zach   McAllister     240      78    R   
24963  floydga01       1983     Gavin        Floyd     245      76    R   

      throws    salary  
24974      R    511300  
24962      L    510900  
24973      R   2500000  
24958      R  13500000  
24957      R   1940000  
24980      R   1500000  
24961      R   2250000  
24967      R   4166666  
24979      L  15000000  
24953      R    507700  
24955      R    900000  
24959      L   5875000  
24956      R   3500000  
24966      L    511900  
24976      R   6000000  
24954      R    547100  
24970      R   6500000  
24981      R    509600  
24971      L   6000000  
24960      R   2337500  
24964      R   1000000  
24968      R   1200000  
24978      R    900000  
24972      R    508600  
24977      R   1550000  
24975      L   2400000  
24965      L    517100  
24969      R    520400  
24963      R   4000000

Suppose that you wanted to study the salaries of all players across all teams that played in the five-year period from 2011 to 2015. First, you can restrict the DataFrame to only the rows involving those years.

In [28]:
years = (players_df['yearid'] >= 2011) & (players_df['yearid'] <= 2015)
five_years = players_df.iloc[years.values, [4, 5, 10, 13]]
five_years.head()

namefirst    namelast  yearid   salary
21454     Henry      Blanco    2011  1000000
21455    Willie  Bloomquist    2011   900000
21456     Geoff        Blum    2011  1350000
21457   Russell     Branyan    2011  1000000
21458       Sam       Demel    2011   417000

Next, create a single column with each player's full name. This way, you can use that column in the next step.

In [29]:
five_years = five_years.assign(fullname = five_years.namefirst + ' ' + five_years.namelast)
five_years.head()

namefirst    namelast  yearid   salary           fullname
21454     Henry      Blanco    2011  1000000       Henry Blanco
21455    Willie  Bloomquist    2011   900000  Willie Bloomquist
21456     Geoff        Blum    2011  1350000         Geoff Blum
21457   Russell     Branyan    2011  1000000    Russell Branyan
21458       Sam       Demel    2011   417000          Sam Demel

At this point, the DataFrame contains the data that you want, but it's not in a format that is easy to make sense of. There is a row for each year that a player played on a team. So each player appears multiple times in the DataFrame. For example, you can look for Henry Blanco:

In [30]:
five_years.iloc[(five_years['fullname'] == 'Henry Blanco').values]

namefirst namelast  yearid   salary      fullname
21454     Henry   Blanco    2011  1000000  Henry Blanco
22293     Henry   Blanco    2012  1250000  Henry Blanco
23902     Henry   Blanco    2013   750000  Henry Blanco

Henry Blanco played in 2011, 2012, and 2013. You can see his salary over those years too. But you want to be able to see this data for all players. It may be possible to sort by the player name, but then comparing salaries for a single year is difficult. What would help is a pivot table—one that's very similar to the pivot tables that you create in a spreadsheet. A pivot table reshapes the DataFrame from a long format to a wide format. You could create a table with the players' names in a single column and each year as a column. Each year column would then contain the salary for that player for that year.

## Grouping 

The `groupby()` method allows you to group records by one or more categorical variables. Typically, you will want to chain some data manipulation method after `groupby()` to summarize the data. 

For example, imagine that you wanted to calculate the average for your statistics, grouped by each year. First, group the DataFrame by `yearid`, then calculate the averages with `mean()`:

In [31]:
players_df.groupby(['yearid']).mean()

birthyear    deathyear      weight     height        salary  \
yearid                                                                  
1985    1954.670909  2007.176471  187.078182  73.298182  4.762994e+05   
1986    1956.552846  2007.657895  187.535230  73.319783  4.171470e+05   
1987    1957.757576  2007.031250  187.397129  73.282297  4.347295e+05   
1988    1958.800905  2007.363636  187.570136  73.352941  4.531711e+05   
1989    1959.907173  2006.818182  188.007032  73.378340  5.063231e+05   
1990    1961.440600  2007.944444  187.876586  73.337947  5.119737e+05   
1991    1961.823358  2008.000000  188.778102  73.426277  8.949612e+05   
1992    1962.712614  2008.419355  188.873862  73.379714  1.047521e+06   
1993    1964.230769  2008.392857  189.726977  73.401950  9.769666e+05   
1994    1965.047511  2010.000000  189.771493  73.450226  1.049589e+06   
1995    1966.190669  2008.950000  190.334686  73.356998  9.649791e+05   
1996    1966.933405  2009.882353  191.357680  73.440387  1.027909e+06   
1997    1967.793514  2010.000000  192.011892  73.430270  1.218687e+06   
1998    1968.765531  2009.000000  193.331663  73.490982  1.280845e+06   
1999    1969.745527  2009.928571  194.197813  73.492048  1.485317e+06   
2000    1970.144737  2010.100000  194.863636  73.435407  1.992985e+06   
2001    1971.012791  2009.000000  195.859302  73.476744  2.279841e+06   
2002    1971.994090  2010.888889  197.478723  73.423168  2.392527e+06   
2003    1973.010883  2010.000000  199.470375  73.465538  2.573473e+06   
2004    1973.845969  2010.000000  200.992780  73.471721  2.491776e+06   
2005    1974.767750  2011.375000  202.623345  73.546330  2.633831e+06   
2006    1975.859585  2011.142857  205.477411  73.548230  2.834521e+06   
2007    1976.994062  2011.000000  207.755344  73.614014  2.941436e+06   
2008    1978.273364  2015.200000  209.542056  73.636682  3.136517e+06   
2009    1979.195572  2013.750000  211.409594  73.697417  3.277647e+06   
2010    1980.243373  2016.333333  212.522892  73.697590  3.278747e+06   
2011    1981.383790  2016.000000  214.241955  73.754470  3.318838e+06   
2012    1982.429245  2016.000000  213.938679  73.737028  3.458421e+06   
2013    1983.611043  2016.000000  213.623313  73.730061  3.723344e+06   
2014    1984.573566  2016.750000  213.018703  73.627182  3.980446e+06   
2015    1985.434517  2016.500000  212.720930  73.649939  4.301276e+06   
2016    1986.712778  2016.500000  212.577960  73.790152  4.396410e+06   

              bmi  
yearid             
1985    24.461572  
1986    24.502401  
1987    24.500915  
1988    24.478833  
1989    24.521215  
1990    24.532075  
1991    24.591785  
1992    24.639914  
1993    24.735772  
1994    24.706733  
1995    24.845307  
1996    24.920078  
1997    25.011439  
1998    25.143549  
1999    25.257313  
2000    25.381477  
2001    25.485376  
2002    25.726705  
2003    25.962863  
2004    26.154659  
2005    26.309108  
2006    26.679922  
2007    26.934415  
2008    27.150872  
2009    27.353613  
2010    27.491214  
2011    27.671684  
2012    27.646131  
2013    27.618469  
2014    27.614355  
2015    27.560812  
2016    27.442015

For another example, you could calculate descriptive statistics for all records, grouped by batting and throwing arms. Try indexing the results to retrieve just the statistics for *salary*.

In [32]:
players_df.groupby(['bats','throws'])['salary'].describe()

count          mean           std      min       25%       50%  \
bats throws                                                                     
B    L         242.0  2.157601e+06  3.521926e+06  60000.0  208125.0  548571.5   
     R        2330.0  2.021790e+06  3.262877e+06  60000.0  300962.5  625000.0   
L    L        4691.0  2.170384e+06  3.644781e+06      0.0  300000.0  560000.0   
     R        2797.0  2.231821e+06  3.722539e+06  60000.0  325000.0  610000.0   
R    L         726.0  1.653911e+06  2.679329e+06  60000.0  230000.0  518600.0   
     R       15642.0  2.062512e+06  3.404014e+06      0.0  275000.0  537500.0   

                   75%         max  
bats throws                         
B    L       2313750.0  15000000.0  
     R       2300000.0  23125000.0  
L    L       2495833.5  33000000.0  
     R       2350000.0  24000000.0  
R    L       1975000.0  16000000.0  
     R       2350000.0  33000000.0

## Reshaping the DataFrame

### The `pivot()` function

The data that you have consists of a list of players and their salaries for the years that they played. To create a pivot table, you need to specify the index; that is, specify which column will be used to identify each row of the DataFrame. Then, specify which column values are going to become the column names of the table. Finally, specify which column will provide the values that go in the table cells.

One important consideration: the index column and column names must be unique in the DataFrame. If a player name and year are duplicated—maybe the player played for two different teams in the same year—you will get an error when trying to create the pivot table. There are many different ways that you can deal with duplicates in the data, and careful thought should be given to this problem. As a quick and simple solution here, you can simply drop all duplicates.

In [33]:
# First, get rid of duplicates that may cause problems
five_years = five_years.drop_duplicates(['fullname', 'yearid'])

# Then create the pivot table
five_years_pivot = five_years.pivot(index='fullname', columns='yearid', values='salary')

print(five_years_pivot)

yearid                  2011        2012        2013        2014        2015
fullname                                                                    
A. J. Burnett     16500000.0  16500000.0  16500000.0  15000000.0   8500000.0
A. J. Ellis         421000.0    490000.0   2000000.0   3550000.0   4250000.0
A. J. Griffin            NaN         NaN    492500.0         NaN         NaN
A. J. Pierzynski   2000000.0   6000000.0   7500000.0   8250000.0   2000000.0
A. J. Pollock            NaN         NaN    491000.0    507000.0    519500.0
...                      ...         ...         ...         ...         ...
Zach Walters             NaN         NaN         NaN         NaN    509600.0
Zach Wheeler             NaN         NaN         NaN    512375.0         NaN
Zack Cozart              NaN    480000.0    497500.0    600000.0   2350000.0
Zack Greinke      13500000.0  13500000.0  21000000.0  26000000.0  25000000.0
Zoilo Almonte            NaN         NaN         NaN    511300.0         NaN

You may notice that the DataFrame now has *yearid* placed on top of *fullname*. This arrangement is known as *hierarchical index* or *multilevel index*. For more on advanced indexing, check the [pandas documentation](https://pandas.pydata.org/pandas-docs/stable/user_guide/advanced.html).

To pivot the data so that it appears in a more familiar tabular format, chain the `reset_index()` method after you `pivot()` the DataFrame:

In [34]:
# Then create the pivot table
five_years_pivot = five_years.pivot(index='fullname', columns='yearid', values='salary').reset_index()

print(five_years_pivot)

yearid          fullname        2011        2012        2013        2014  \
0          A. J. Burnett  16500000.0  16500000.0  16500000.0  15000000.0   
1            A. J. Ellis    421000.0    490000.0   2000000.0   3550000.0   
2          A. J. Griffin         NaN         NaN    492500.0         NaN   
3       A. J. Pierzynski   2000000.0   6000000.0   7500000.0   8250000.0   
4          A. J. Pollock         NaN         NaN    491000.0    507000.0   
...                  ...         ...         ...         ...         ...   
1517        Zach Walters         NaN         NaN         NaN         NaN   
1518        Zach Wheeler         NaN         NaN         NaN    512375.0   
1519         Zack Cozart         NaN    480000.0    497500.0    600000.0   
1520        Zack Greinke  13500000.0  13500000.0  21000000.0  26000000.0   
1521       Zoilo Almonte         NaN         NaN         NaN    511300.0   

yearid        2015  
0        8500000.0  
1        4250000.0  
2              NaN  
3  

### The `melt()` function

With the `melt()` function, you can reshape a DataFrame. 

This function takes many arguments, following this form:

`pandas.melt(frame, id_vars=None, value_vars=None, var_name=None, value_name='value', col_level=None)`


Here's what the first five arguments do, according to [pandas documentation](https://www.geeksforgeeks.org/python-pandas-melt/):


| Argument     | What it does                                                                               |
| ------------ | ------------------------------------------------------------------------------------------ |
| `frame`    | DataFrame                                                    |
| `id_vars`    | Column or columns to use as identifier columns                                                    |
| `value_vars` | Column or columns to unpivot. If not specified, uses all columns that are not set as `id_vars`.    |
| `var_name`   | Name to use for the `variable` column. If none, it uses `frame.columns.name` or `variable`. |
| `value_name` | Name to use for the `value` column.                                                        |
| `col_level`  | If columns are a MultiIndex, then use this level to melt.                                   |

Using `melt()`, how can you unpivot the *five_years_pivot* table from above? Rather than having five columns for `2011`-`2015` indicating salary information, you just want one column called *year* and another called *salary*. 

That means that *fullname* is your `id_vars`, and the rest of the columns are your `value_vars`. You'll set `var_name` to *year* and `value_name` to *salary*, as previously specified. 



In [35]:
pd.melt(frame=five_years_pivot, id_vars=['fullname'],var_name='year',value_name='salary')

fullname  year      salary
0        A. J. Burnett  2011  16500000.0
1          A. J. Ellis  2011    421000.0
2        A. J. Griffin  2011         NaN
3     A. J. Pierzynski  2011   2000000.0
4        A. J. Pollock  2011         NaN
...                ...   ...         ...
7605      Zach Walters  2015    509600.0
7606      Zach Wheeler  2015         NaN
7607       Zack Cozart  2015   2350000.0
7608      Zack Greinke  2015  25000000.0
7609     Zoilo Almonte  2015         NaN

[7610 rows x 3 columns]

## Merging DataFrames with the `merge()` function

Joining two or more tables is one of the most common data preparation tasks, and there are many ways to do it. You will be introduced to the world of joins later in the program with SQL. This will give you a solid grasp of how joins are conducted in pandas, because pandas joins are modeled off of SQL joins. 

For now, fortunately, you can rely on the intuition of pandas to give you what you want when joining two tables.

The above DataFrame consisting of player information and salaries is actually the result of joining two individual DataFrames. You can reproduce this DataFrame using `merge()`. For now, call the `merge()` function and pass in the two DataFrames, assigning the results to a new DataFrame.

Pandas will make a best guess on how to merge these DataFrames; if any two columns are named the same between them, that will serve as the basis for the join. In this case, the *playerID* field is shared between the two tables, so pandas will join based on that connection.  





In [36]:
salaries = pd.read_csv("https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/salaries.csv")
people = pd.read_csv("https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/people.csv")

print(salaries.columns)
print(people.columns)

Index(['yearid', 'teamid', 'lgid', 'playerid', 'salary'], dtype='object')
Index(['playerid', 'birthyear', 'birthcountry', 'deathyear', 'namefirst',
       'namelast', 'weight', 'height', 'bats', 'throws'],
      dtype='object')


In [37]:
merged = pd.merge(salaries, people)

print(merged.columns)
print(merged.shape)

Index(['yearid', 'teamid', 'lgid', 'playerid', 'salary', 'birthyear',
       'birthcountry', 'deathyear', 'namefirst', 'namelast', 'weight',
       'height', 'bats', 'throws'],
      dtype='object')
(26428, 14)


Indeed, this is the same information as contained in the original DataFrame, *players_df*.

In [38]:
players_df = pd.read_csv("https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/players.csv")

print(players_df.columns)
print(players_df.shape)

Index(['playerid', 'birthyear', 'birthcountry', 'deathyear', 'namefirst',
       'namelast', 'weight', 'height', 'bats', 'throws', 'yearid', 'teamid',
       'lgid', 'salary'],
      dtype='object')
(26428, 14)


Should you need to perform another type of join, such as one when columns do not have the same name, you can check the [pandas documentation](https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html). However, as mentioned earlier, the upcoming lessons on SQL will explore the logic behind how joins are conducted.

## Case study 1

Your manager would like to study the distribution of right-handed batters to right-handed throwers. They want to know how many right-handed batters are also right-handed throwers, and they are also interested in other combinations. You look at the original data and realize that you will have to perform some sort of count over all the rows where the value for bats is `right-handed` and the value for throws is `right-handed`, and you'll do something similar for other combinations.

The data contains duplicate rows for each player, because a row actually represents a single year per player. You do not need all of that information, so you could start by dropping the duplicate players.

In [39]:
players_no_duplicates = players_df.drop_duplicates(['playerid'])
players_no_duplicates.shape

(5149, 14)

That leaves 5,149 players. Your manager wants a table of data that shows the counts of each combination of bats and throws. This is an ideal scenario for a *crosstab*. To create a crosstab, simply provide the two columns of data that you wish to cross-tabulate, and pandas will do the rest for you.

In [40]:
pd.crosstab(players_no_duplicates['bats'], players_no_duplicates['throws'])

throws    L     R
bats             
B        44   400
L       918   520
R       152  3115

## Case study 2

Your company got some data about the fuel use of a number of vehicles in a relational database named *fueleconomy*. This database is on the same server as the previous database. You would like to answer some questions about this data, such as the following:

1. How many vehicles use each type of fuel by year?
2. What are the top-performing six-cylinder vehicles in 2003, as measured by highway mileage?


First, connect to the database and get the data into a DataFrame.


In [41]:
vehicles_df = pd.read_csv("https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/vehicles.csv")

Next, look at some features of the data that you got. For example, how many rows are there? What are the data types?

In [42]:
vehicles_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33442 entries, 0 to 33441
Data columns (total 12 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   id      33442 non-null  int64  
 1   make    33442 non-null  object 
 2   model   33442 non-null  object 
 3   year    33442 non-null  int64  
 4   class   33442 non-null  object 
 5   trans   33434 non-null  object 
 6   drive   33442 non-null  object 
 7   cyl     33384 non-null  float64
 8   displ   33385 non-null  float64
 9   fuel    33442 non-null  object 
 10  hwy     33442 non-null  int64  
 11  cty     33442 non-null  int64  
dtypes: float64(2), int64(4), object(6)
memory usage: 3.1+ MB


Next, you can check how many years of data are contained in the DataFrame. That is, how many unique years are there?

In [43]:
vehicles_df['year'].value_counts()

1985    1701
1987    1247
2014    1214
1986    1210
2008    1186
2013    1184
2009    1184
2005    1166
1989    1153
2012    1153
1991    1132
2011    1130
1988    1130
2007    1126
2004    1122
1992    1121
2010    1109
2006    1104
1993    1093
1990    1078
2003    1044
1994     982
2002     974
1995     967
2001     909
1999     849
2000     839
1998     811
1984     784
1996     773
1997     762
2015     205
Name: year, dtype: int64

Looks like you have data from 1985 to 2015. How many different types of fuel are there?

In [44]:
vehicles_df['fuel'].value_counts()

Regular                        22622
Premium                         8617
Gasoline or E85                 1043
Diesel                           874
Premium or E85                    88
CNG                               58
Electricity                       55
Midgrade                          43
Gasoline or natural gas           18
Regular Gas and Electricity        8
Gasoline or propane                8
Premium Gas or Electricity         7
Premium and Electricity            1
Name: fuel, dtype: int64

Then do a crosstab between these two columns and see what you learn.

In [45]:
pd.crosstab(vehicles_df['year'], vehicles_df['fuel'])

fuel  CNG  Diesel  Electricity  Gasoline or E85  Gasoline or natural gas  \
year                                                                       
1984    0     105            0                0                        0   
1985    0     158            0                0                        0   
1986    0      71            0                0                        0   
1987    0      56            0                0                        0   
1988    0      31            0                0                        0   
1989    0      33            0                0                        0   
1990    0      29            0                0                        0   
1991    0      39            0                0                        0   
1992    0      23            0                0                        0   
1993    1      22            0                0                        0   
1994    1      31            0                0                        0   
1995    1      32            0                0                        0   
1996    1      13            0                0                        0   
1997    1      11            0                0                        0   
1998    3      14            2                0                        0   
1999    5      10            4                0                        0   
2000    5       6            3               14                        1   
2001    7       6            3               18                        3   
2002    5       8            2               27                        3   
2003    4       9            1               35                        4   
2004    8      11            0               37                        7   
2005    5      14            0               40                        0   
2006    1       9            0               33                        0   
2007    1       7            0               56                        0   
2008    1       7            0               60                        0   
2009    1      12            0               91                        0   
2010    1      14            0               96                        0   
2011    2      16            4              137                        0   
2012    2      17            9              140                        0   
2013    1      22           14              135                        0   
2014    1      35           12              111                        0   
2015    0       3            1               13                        0   

fuel  Gasoline or propane  Midgrade  Premium  Premium Gas or Electricity  \
year                                                                       
1984                    0         0        0                           0   
1985                    0         0      101                           0   
1986                    0         0       75                           0   
1987                    0         0       89                           0   
1988                    0         0      104                           0   
1989                    0         0      142                           0   
1990                    0         0      128                           0   
1991                    0         0      198                           0   
1992                    0         0      221                           0   
1993                    0         0      209                           0   
1994                    0         0      204                           0   
1995                    0         0      201                           0   
1996                    0         0      147                           0   
1997                    0         0      169                           0   
1998                    0         0      155                           0   
1999                    0         0      174                           0   
2000                    0         0      234   

That gives you a useful look at the way that different fuels have been used over the years. In an upcoming lesson, you will learn how to make charts out of this type of data. 

The second question will require a few steps. First, select only the six-cylinder vehicles from 2003.

In [46]:
bools = (vehicles_df['cyl'] == 6) & (vehicles_df['year'] == 2003)
six_cyl_df = vehicles_df.iloc[bools.values]
six_cyl_df.head()

id   make    model  year                        class  \
25   18458  Acura    3.2CL  2003                 Compact Cars   
26   18459  Acura    3.2CL  2003                 Compact Cars   
31   18629  Acura    3.2TL  2003                 Midsize Cars   
39   18630  Acura    3.5RL  2003                 Midsize Cars   
120  19118  Acura  MDX 4WD  2003  Sport Utility Vehicle - 4WD   

               trans                       drive  cyl  displ     fuel  hwy  \
25      Manual 6-spd           Front-Wheel Drive  6.0    3.2  Premium   26   
26    Automatic (S5)           Front-Wheel Drive  6.0    3.2  Premium   27   
31    Automatic (S5)           Front-Wheel Drive  6.0    3.2  Premium   27   
39   Automatic 4-spd           Front-Wheel Drive  6.0    3.5  Premium   22   
120  Automatic 5-spd  4-Wheel or All-Wheel Drive  6.0    3.5  Premium   21   

     cty  
25    17  
26    17  
31    17  
39    16  
120   15

Next, you can either sort the entire DataFrame by city mileage in descending order, or use the built-in `nlargest()` method to select the *n* rows with the largest values in the highway column. You can look at both solutions just to compare approaches.

In [47]:
# Sort the DataFrame by 'hwy' in descending order
sorted_six_cyl_df = six_cyl_df.sort_values(by='hwy', ascending=False)

# Select the first 10 rows
sorted_six_cyl_df.head(10)

id        make        model  year            class            trans  \
5450   18731   Chevrolet       Impala  2003       Large Cars  Automatic 4-spd   
5873   18646   Chevrolet  Monte Carlo  2003     Midsize Cars  Automatic 4-spd   
32123  18619  Volkswagen        Jetta  2003     Compact Cars   Automatic (S5)   
1403   18469         BMW         325i  2003     Compact Cars     Manual 5-spd   
1637   18473         BMW         330i  2003     Compact Cars     Manual 5-spd   
3056   18641       Buick        Regal  2003     Midsize Cars  Automatic 4-spd   
2984   18727       Buick  Park Avenue  2003       Large Cars  Automatic 4-spd   
15922  18658       Honda       Accord  2003     Midsize Cars     Manual 6-spd   
15921  18657       Honda       Accord  2003     Midsize Cars  Automatic 5-spd   
1609   18400         BMW        330ci  2003  Subcompact Cars     Manual 5-spd   

                   drive  cyl  displ     fuel  hwy  cty  
5450   Front-Wheel Drive  6.0    3.4  Regular   29   19  
5873   Front-Wheel Drive  6.0    3.4  Regular   29   19  
32123  Front-Wheel Drive  6.0    2.8  Premium   28   19  
1403    Rear-Wheel Drive  6.0    2.5  Premium   27   18  
1637    Rear-Wheel Drive  6.0    3.0  Premium   27   18  
3056   Front-Wheel Drive  6.0    3.8  Regular   27   17  
2984   Front-Wheel Drive  6.0    3.8  Regular   27   17  
15922  Front-Wheel Drive  6.0    3.0  Regular   27   18  
15921  Front-Wheel Drive  6.0    3.0  Regular   27   19  
1609    Rear-Wheel Drive  6.0    3.0  Premium   27   18

That did the trick. You have the top ten vehicles by highway mileage in 2003. Notice that there are several vehicles with the same mileage values. Now, see how  `nlargest` may work to do the same job.

`nlargest` is really just a shortcut. You need to provide the following:

 - `n`: The number of rows to return
 - `columns`: The columns to sort by
 - `keep`: One of `first`, `last`, or `all`. This affects how you deal with duplicate values in the sort column, and it defaults to `first`.

In [48]:
# Select the 10 rows with largest values in the 'hwy' column
six_cyl_df.nlargest(10, columns='hwy')

id        make             model  year                 class  \
5450   18731   Chevrolet            Impala  2003            Large Cars   
5873   18646   Chevrolet       Monte Carlo  2003          Midsize Cars   
32123  18619  Volkswagen             Jetta  2003          Compact Cars   
26     18459       Acura             3.2CL  2003          Compact Cars   
31     18629       Acura             3.2TL  2003          Midsize Cars   
1376   18398         BMW             325ci  2003       Subcompact Cars   
1403   18469         BMW              325i  2003          Compact Cars   
1430   18761         BMW  325i Sport Wagon  2003  Small Station Wagons   
1609   18400         BMW             330ci  2003       Subcompact Cars   
1637   18473         BMW              330i  2003          Compact Cars   

                 trans              drive  cyl  displ     fuel  hwy  cty  
5450   Automatic 4-spd  Front-Wheel Drive  6.0    3.4  Regular   29   19  
5873   Automatic 4-spd  Front-Wheel Drive  6.0    3.4  Regular   29   19  
32123   Automatic (S5)  Front-Wheel Drive  6.0    2.8  Premium   28   19  
26      Automatic (S5)  Front-Wheel Drive  6.0    3.2  Premium   27   17  
31      Automatic (S5)  Front-Wheel Drive  6.0    3.2  Premium   27   17  
1376      Manual 5-spd   Rear-Wheel Drive  6.0    2.5  Premium   27   18  
1403      Manual 5-spd   Rear-Wheel Drive  6.0    2.5  Premium   27   18  
1430      Manual 5-spd   Rear-Wheel Drive  6.0    2.5  Premium   27   18  
1609      Manual 5-spd   Rear-Wheel Drive  6.0    3.0  Premium   27   18  
1637      Manual 5-spd   Rear-Wheel Drive  6.0    3.0  Premium   27   18

The exact rows that are returned may not exactly match because of the way that `nlargest` deals with duplicates. By default, it returns the first rows in the set of duplicates. Experiment with this feature by trying `last` and `all` as values to the `keep` parameter.